File: src/memento/History.ts
import { CalculatorMemento } from "./CalculatorMemento";
import { Calculator } from "./Calculator";

export class History {
    private undoStack: CalculatorMemento[] = [];
    private redoStack: CalculatorMemento[] = [];

    save(memento: CalculatorMemento): void {
        this.undoStack.push(memento);
        this.redoStack = [];
    }

    undo(calculator: Calculator): CalculatorMemento | undefined {
        if (this.undoStack.length > 0) {
            this.redoStack.push(calculator.save());
            return this.undoStack.pop();
        }
        return undefined;
    }

    redo(calculator: Calculator): CalculatorMemento | undefined {
        if (this.redoStack.length > 0) {
            this.undoStack.push(calculator.save());
            return this.redoStack.pop();
        }
        return undefined;
    }

    canUndo(): boolean {
        return this.undoStack.length > 0;
    }

    canRedo(): boolean {
        return this.redoStack.length > 0;
    }
}

File: src/main/java/calculadora/memento/History.java
package calculadora.memento;

/**
 *
 * @author vedz11
 */
import java.util.Stack;

public class History {
    private final Stack<CalculatorMemento> undoStack = new Stack<>();
    private final Stack<CalculatorMemento> redoStack = new Stack<>();

    public void save(CalculatorMemento memento) {
        undoStack.push(memento);
        redoStack.clear(); // limpiamos el redo al hacer nueva operación
    }

    public CalculatorMemento undo(Calculator calculator) {
        if (!undoStack.isEmpty()) {
            // Guardar el estado actual para posible redo
            redoStack.push(calculator.save());
            return undoStack.pop();
        }
        return null;
    }

    public CalculatorMemento redo(Calculator calculator) {
        if (!redoStack.isEmpty()) {
            // Guardar el estado actual para posible undo
            undoStack.push(calculator.save());
            return redoStack.pop();
        }
        return null;
    }

    public boolean canUndo() {
        return !undoStack.isEmpty();
    }

    public boolean canRedo() {
        return !redoStack.isEmpty();
    }
}
